File: src/java/ordination/QuickSort.java
package ordination;

public class QuickSort {

    public static void exchange(int v[], int start, int end) {
        int aux;

        aux = v[start];
        v[start] = v[end];
        v[end] = aux;
    }

    public static int separate(int v[], int start, int end) {
        int pivot, homePosition, finalPosition;

        pivot = v[start];
        homePosition = start - 1;
        finalPosition = end + 1;

        while (true) {
            do {
                --finalPosition;
            } while (v[finalPosition] > pivot);

            do {
                homePosition++;
            } while (v[homePosition] < pivot);

            if (homePosition < finalPosition) {
                exchange(v, homePosition, finalPosition);
            } else {
                return finalPosition;
            }
        }
    }

    public static void quickSort(int start, int end, int v[]) {

        while (start < end) {
            int q = separate(v, start, end);
            quickSort(start, q, v);
            start = q + 1;
        }
    }

    public static void quickSort2(int start, int end, int v[]) {

        int kinda;

        if (start < end) {
            kinda = separate(v, start, end);
            quickSort2(start, kinda, v);
            quickSort2(kinda + 1, end, v);
        }
    }

}

File: src/c++/ordination/QuickSort.cpp
#include <stdio.h>
#define MAX 100

void exchange(int v[MAX], int p, int r) {
    int aux;

    aux = v[p];
    v[p] = v[r];
    v[r] = aux;
}

int separa(int p, int r, int v[MAX]) {
    int pivot, i, j;

    pivot = v[p];
    i = p - 1;
    j = r + 1;

    while (true) {

        do {
            --j;
        } while(v[j] > pivot);

        do {
            i++;
        } while(v[i] < pivot);

        if (i < j) {
            exchange(v, i, j);
        } else {
            return j;
        }
    }
}

int partition(int p, int r, int v[MAX]) {
	int x = v[r];
	
	int i = p - 1;
	
	for (int j = p; j < r; j++) {
        if (v[j] <= x) {
			i++;
			exchange(v, v[j], v[i]);
		}
	}
	
	exchange(v, v[i + 1], v[r]);
	return i + 1;
}

void quickSort(int p, int r, int v[MAX])
{
	if (p < r)
	{
		int q = partition(p, r, v);
		
		quickSort(p, q - 1, v);
		quickSort(q + 1, r, v);
	}
}